### PR TITLE
TINKERPOP-2095 GroupStep looks for irrelevant barrier steps

### PR DESCRIPTION
**Pull request by dkuppitz**
https://issues.apache.org/jira/browse/TINKERPOP-2095

This PR fixes a bug that somehow managed to stay unnoticed for years.

Prior this PR we got this:

```
gremlin> g.V().hasLabel('person').as('p').out('created').group().by('name').by(select('p').values('age').sum())
java.lang.Long cannot be cast to org.apache.tinkerpop.gremlin.process.traversal.traverser.util.TraverserSet
```

Now we get this:

```
gremlin> g.V().hasLabel('person').as('p').out('created').group().by('name').by(select('p').values('age').sum())
==>[ripple:32,lop:96]
```

The fix was made for both - `GroupStep` and `GroupSideEffectStep`.

I also had to fix `ComputerVerificationStrategy`; it didn't detect that `select('p').values('age')` is leaving the star-graph. The actual [fix](https://github.com/apache/tinkerpop/pull/997/files#diff-85643f456bbefa4cd579cd5ef1db9410R132) is in `TraversalHelper::isLocalStarGraph`, but this method is mainly used by `ComputerVerificationStrategy`.

Without the `ComputerVerificationStrategy` fix:

```
gremlin> g.V().hasLabel('person').as('p').out('created').group().by('name').by(select('p').values('age').sum())
==>[ripple:0,lop:0]
```

With the `ComputerVerificationStrategy` fix:

```
gremlin> g.V().hasLabel('person').as('p').out('created').group().by('name').by(select('p').values('age').sum())
Local traversals may not traverse past the local star-graph on GraphComputer: [SelectOneStep(last,p), PropertiesStep([age],value), SumGlobalStep]
```

`docker/build.sh -t -i -n` passed.

VOTE +1

### CODE REVIEW

**Comment by spmallette** [comment]
crazy one - VOTE +1

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/GroupStep.java [hunk 1]
```diff
@@ -59,19 +60,36 @@
...
 
+    /**
+     * Determines the first (non-local) barrier step in the provided traversal. This method is used by {@link GroupStep}
+     * and {@link org.apache.tinkerpop.gremlin.process.traversal.step.sideEffect.GroupSideEffectStep} to ultimately
+     * determine the reducing reducing bi-operator.
+     *
+     * @param traversal The traversal to inspect.
+     * @return The first non-local barrier step or {@code null} if no such step was found.
+     */
+    public static <S, V> Barrier determineBarrierStep(final Traversal.Admin<S, V> traversal) {
```
Comment (jorgebay) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/GroupStep.java]: Do we have to expose this method or can it be a package private method?

Comment (dkuppitz) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/GroupStep.java]: I wanted to make it package private, but unfortunately `GroupSideEffectStep` resides in a different package.

Comment (jorgebay) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/GroupStep.java]: Makes sense.

**Pull request merged**